cmd/clusterctl/pkg/client/config/reader_viper.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/viper"
+	"k8s.io/client-go/util/homedir"
+	"k8s.io/klog"
+)
+
+// viperReader implements Reader using viper as backend for reading from environment variables
+// and from a clusterctl config file.
+type viperReader struct {
+}
+
+// newViperReader returns a viperReader.
+func newViperReader() Reader {
+	return &viperReader{}
+}
+
+// Init initialize the viperReader.
+func (v *viperReader) Init(path string) error {
+	if path != "" {
+		// Use path file from the flag.
+		viper.SetConfigFile(path)
+	} else {
+		// Configure for searching cluster-api/.clusterctl{.extension} in home directory
+		viper.SetConfigName(".clusterctl")
+		viper.AddConfigPath(filepath.Join(homedir.HomeDir(), "cluster-api"))
+	}
+
+	// Configure for reading environment variables as well (with - to _ replacement in the variable name)
+	replacer := strings.NewReplacer("-", "_")
+	viper.SetEnvKeyReplacer(replacer)
+
+	viper.AutomaticEnv()
```

Comment:
Add a comment describing what does this do and why we're using it

Reply:
done